### PR TITLE
ARROW-12000: [Documentation] Add note about deviation from style guide on struct/classes

### PR DESCRIPTION
**Pull request by westonpace**
docs

### CODE REVIEW

File: docs/source/developers/cpp/development.rst [hunk 1]
```diff
@@ -77,6 +77,10 @@ This project follows `Google's C++ Style Guide
 * We relax the line length restriction to 90 characters.
 * We use the ``NULLPTR`` macro in header files (instead of ``nullptr``) defined
   in ``src/arrow/util/macros.h`` to support building C++/CLI (ARROW-1134)
+* For internal types (not in public headers) it is ok to use structs for
+  convenience (e.g. list initialization) if access control is not needed
+  (e.g. no public access to a nested type) even if the struct has
+  state/invaiants
```
Comment (dianaclarke) [inline docs/source/developers/cpp/development.rst:83]: `s/invaiants/invariants/` ?

**Comment by pitrou** [comment]
This doesn't seem exact. We have `struct` types in public headers (such as `filesystem.h`).
I think @wesm 's rationale seems more faithful to reality:
> In the public headers we should be trying to use struct only for objects that are principally simple data containers where we are OK with exposing all the internal members, and any methods are primarily conveniences.

**Comment by westonpace** [comment]
FileInfo is an interesting case.  It actually does use access control and has mutable state.  Wes' rationale covers the public headers case but I also want to say something about private headers.  I'll try again.

**Comment by westonpace** [comment]
docs2

@pitrou Does this wording seem better?

**Comment by pitrou** [comment]
Yes, it reads good to me.

**Pull request closed without merging**